server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -1437,11 +1460,28 @@ public ClusterState getClusterStateUsingDiff(ClusterMetadataManifest manifest, C
             indexRoutingTables.remove(indexName);
         }
 
-        return clusterStateBuilder.stateUUID(manifest.getStateUUID())
+        ClusterState clusterState = clusterStateBuilder.stateUUID(manifest.getStateUUID())
             .version(manifest.getStateVersion())
             .metadata(metadataBuilder)
             .routingTable(new RoutingTable(manifest.getRoutingTableVersion(), indexRoutingTables))
             .build();
+
+        if (checksumValidationEnabled && manifest.getClusterStateChecksum()!=null) {
+            validateClusterStateFromChecksum(manifest.getClusterStateChecksum(), clusterState);
+        }
+        return clusterState;
+    }
+
+    private void validateClusterStateFromChecksum(ClusterStateChecksum clusterStateChecksum, ClusterState clusterState) {
+        ClusterStateChecksum newClusterStateChecksum = new ClusterStateChecksum(clusterState);
+        if (!newClusterStateChecksum.equals(clusterStateChecksum)) {
+            logger.error(
+                "Cluster state checksums do not match. Checksum from manifest {}, checksum from created cluster state {}",
+                clusterStateChecksum,
+                newClusterStateChecksum
+            );
+            throw new IllegalStateException("Cluster state checksums do not match.");
```

Comment:
nit add expected vs actual and also which sub-component's checksum failed

Reply:
added